### PR TITLE
Buchanan, Allison done besides plots function

### PR DESCRIPTION
**Pull request by allisonw5**
# Pull Request Template

### Reproducibility
- [x] Are the data downloaded in the code
- [x] Are paths created to ensure they work on all operating systems (using  os.path.join)
- [x] Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
- [x] Does the notebook run from start to finish?

### Functions
- [x] Do functions follow PEP 8 format conventions?
- [x] Are function docstrings clear (all  inputs and outputs clearly  described and defined)
- [x] Are function names expressive (the name describes what the function does)?
- [x] Are functions easy to understand and read?
- [x] How many tasks does each function do? (ideally a function does one thing well).

### Package imports
- [x] Are standard modules (those included with the base  python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
- [x] Are variable names throughout the code, expressive?
- [x] Suggest changes if not, highlight what was done well
- [x] Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?

### DRY Code
- [x] Are segments of code repeated in the file or is the code DRY?
- [x] Are loops used to optimize DRY code?
- [x] Are functions used to optimize DRY code?
- [x] Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could  be cleaner / more efficient.

### Novel Approaches to Problem solving
- [x] Highlight any novel approaches to completing the assignment.

### CODE REVIEW

**Comment by eculler** [comment]
@jen-morse Please review this PR #57  (Canvas isn't working so let's try GitHub directly!)

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 1]
```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # LiDAR Uncertainty: A Comparison of Two California Sites.
+
+# ## The Soap Site
+# ![image.png](attachment:image.png)
+# image credit: National Ecological Observation Network
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:8]: Hi @allisonw5, small suggestion here: NEON is National Ecological **Observatory** Network (as opposed to Observation)

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+# In[13]:
+
+
+## The SJER Site
+get_ipython().system('[image.png](attachment:image.png)')
+image credit: National Ecological Observation Network
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:15]: same comment as above (observation should be observatory)

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 3]
```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # LiDAR Uncertainty: A Comparison of Two California Sites.
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:4]: May want to be more descriptive here? (e.g., a comparison of LiDAR vs insitu data sets using two forested California Sites, or something like that).

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+# In[91]:
+
+
+class AnalyzeLidarUncertainty:
+
+    '''This class calculates stats for lidar and insitu and merges the two data frames.
+
+        parameters:
+        ----------
+        lidar_path: str (or path?)
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:119]: Remember to update or get rid of question mark for turning in final code. :)

File: AllisonClass [hunk 5]
```diff
@@ -0,0 +1,128 @@
+class NEONDataLoader:
+
+    '''This class creates all of our necessary paths for downloads
+
+    parameters:
```
Comment (Jen-Morse) [inline AllisonClass:5]: For function docstrings I think you want to capitalize 'Return' and 'Parameters'?

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+# Import necessary packages
+import os
+import pathlib
+
+import earthpy as et
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:29]: Elsa mentioned in class at one point it can be good practice to load pandas before geopandas and xarray before rioxarray to avoid dependency issues? (Even though it's alphabetically out of order). You may already know this and know there aren't any issues. But thought I'd mention it! :)

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+soap_data_loader=NEONDataLoader('SOAP', 'ID', 
+                                {'separator':'-', 'plot':''})
+
+soap_data_loader.insitu_path
+
+
+# In[91]:
+
+
+class AnalyzeLidarUncertainty:
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:113]: I like how you divided this class into two, one to 'load' the data and one to 'run' the data. I think that could help with reproducibility with future data sets that require different steps to load, but once loaded could be run in the same way to pull out the max, means. Especially if you included some work here to standardize the column  and or file names?

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                               'mean': 'insitu_mean',
+                               'ID': 'plotid',
+                               'Plot_ID': 'plotid'}))
+
+        return df
+
+    def merge_dfs(self):
+        '''
+        Merges insitu and lidar data frames together.
+
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:186]: extra blank line here

File: Buchanan-Allison-LiDAR-Uncertainty-Class (1).py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+# In[88]:
+
+
+sjer_site.merge_dfs()
+
+
+# In[90]:
+
+
+def plot_regression(lidar_mean, insitu_mean, lidar_max, insitu_max):
```
Comment (Jen-Morse) [inline Buchanan-Allison-LiDAR-Uncertainty-Class (1).py:243]: the code in the .py file for this section is different than the code in the .ipynb file... I'm not good enough yet at python to 'eyeball' which is better w/out running it. But... if you let me know which you intended to be the final version, happy to add more comments or help with any debugging. For either code, it looks like you could easily add site name as a parameter if you wanted to quickly get labels onto your plots. (Though there may be a more efficient method).

**Comment by Jen-Morse** [comment]
Hi @allisonw5!  I hope my comments are helpful to you!
I wan't able to edit the review checklist, so cutting and pasting it below:

Review Checklist

Reproducibility

 Are the data downloaded in the code 
 Are paths created to ensure they work on all operating systems (using os.path.join) 
 Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
Does the notebook run from start to finish?
PEP 8 standards & Code Readability

_Yes to all of the above w/ the exception of a few pep 8 standards which I commented on in the .py file. (Hopefully they are helpful, not sure if you are already aware of giving the code a once over before turning it in)._ 

Functions

 Do functions follow PEP 8 format conventions?
 Are function docstrings clear (all inputs and outputs clearly described and defined)
 Are function names expressive (the name describes what the function does)?
 Are functions easy to understand and read?
 How many tasks does each function do? (ideally a function does one thing well).

_Yes with the same comments a few functions need some words capitalized (I included in comments)._ 

Package imports

 Are standard modules (those included with the base python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
 Are variable names throughout the code, expressive?
 Suggest changes if not, highlight what was done well
 Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?

_Overall the code is great and does what it is supposed to do, is reproducible and the notebook runs from start to finish! Nice work Allison. I think you are still working on the plotting function?  I'm not great w/ loops/functions but a quick way to get the titles for each plot could be to add the site name as a parameter in the plotting function? (Or at least I think that is how I did that, I'm not sure if it's the best way, but a quick way to get those labels on the plots).  As far as making the code more clear; the data loader function I think could potentially use more precise/clear docstrings for what the class does and what it returns. I also made a comment on ordering of libraries which may/may not be relevant for you._ 

DRY Code

 Are segments of code repeated in the file or is the code DRY?
 Are loops used to optimize DRY code?
 Are functions used to optimize DRY code?
 Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could be cleaner / more efficient.

_Yes to all including you use of functions and classes to keep code DRY._

Novel Approaches to Problem solving

 Highlight any novel approaches to completing the assignment.

  _I added this comment as well, that I liked how you separated out a 'load data' class from a 'run data' class, which I think could be useful for using this code w/ additional data sets that do not have standardized column names etc.._

**Review by Jen-Morse — COMMENTED** [review]
Comments added to .py file. Nice work! I did notice the .ipynb file and .py file are not quite the same version. If you wanted to check and see which version you were intending to include for both, I'm happy to add/update any review comments. :)

**Comment by eculler** [comment]
Awesome classes, Allison! This is super-impressive work.

Only quibble - it looks like you are confusing attributes and parameters in your class docstring. Keep an eye on that numpy docstring styleguide.
